File: multimodal-generation/video-generation/javascript/01_text_to_video_generation.js
#!/usr/bin/env node

import { 
    BedrockRuntimeClient, 
    StartAsyncInvokeCommand, 
    GetAsyncInvokeCommand, 
    ListAsyncInvokesCommand 
} from '@aws-sdk/client-bedrock-runtime';
import { S3Client, CreateBucketCommand } from '@aws-sdk/client-s3';
import winston from 'winston';
import {monitorAndDownloadVideo, monitorAndDownloadVideos, saveInvocationInfo} from './amazonVideoUtil.js';

// Configure logging
const logger = winston.createLogger({
    level: 'info',
    format: winston.format.simple(),
    transports: [
        new winston.transports.Console()
    ]
});

// Constants
const DEFAULT_MODEL_ID = "amazon.nova-reel-v1:0";

// Initialize AWS clients
let bedrockRuntime;
let s3Client;

/**
 * Initializes AWS Bedrock Runtime and S3 clients.
 * @param {string} [region="us-east-1"] - The AWS region to initialize clients for.
 */
function setupAwsSession(region = "us-east-1") {
    bedrockRuntime = new BedrockRuntimeClient({ region });
    s3Client = new S3Client({ region });
    logger.info("AWS SDK clients have been initialized.");
}


/**
 * Generates a video using the specified prompt and model.
 * @param {string} s3DestinationBucket - The S3 bucket to store the generated video.
 * @param {string} videoPrompt - The prompt for video generation.
 * @param {string} [modelId=DEFAULT_MODEL_ID] - The model ID to use for video generation.
 * @returns {Promise<string|null>} - The ARN of the invocation if successful, otherwise null.
 */
async function generateVideo(s3DestinationBucket, videoPrompt, modelId = DEFAULT_MODEL_ID) {
    try {
        // Create the S3 bucket
        await s3Client.send(new CreateBucketCommand({ 
            Bucket: s3DestinationBucket 
        }));

        const modelInput = {
            taskType: "TEXT_VIDEO",
            textToVideoParams: {
                text: videoPrompt,
            },
            videoGenerationConfig: {
                durationSeconds: 6,
                fps: 24,
                dimension: "1280x720",
                seed: Math.floor(Math.random() * 2147483648),
            },
        };

        const command = new StartAsyncInvokeCommand({
            modelId: modelId,
            contentType: "application/json",
            accept: "application/json",
            modelInput: modelInput,
            outputDataConfig: {
                s3OutputDataConfig: {
                    s3Uri: `s3://${s3DestinationBucket}`
                }
            }
        });

        const invocation = await bedrockRuntime.send(command);
        const invocationArn = invocation.invocationArn;

        logger.info("\nResponse:");
        logger.info(JSON.stringify(invocation, null, 2));

        await saveInvocationInfo(invocation, modelInput);

        return invocationArn;

    } catch (error) {
        logger.error(error);
        return null;
    }
}

/**
 * Checks the status of a video generation job.
 * @param {string} invocationArn - The ARN of the invocation.
 * @returns {Promise<Object|null>} - The response from the GetAsyncInvokeCommand if successful, otherwise null.
 */
async function checkJobStatus(invocationArn) {
    try {
        const command = new GetAsyncInvokeCommand({ invocationArn });
        const response = await bedrockRuntime.send(command);
        
        const status = response.status;
        logger.info(`Status: ${status}`);
        logger.info("\nFull response:");
        logger.info(JSON.stringify(response, null, 2));
        return response;
    } catch (error) {
        logger.error(`Error checking job status: ${error}`);
        return null;
    }
}

/**
 * Lists the statuses of video generation jobs.
 * @param {number} [maxResults=10] - The maximum number of results to return.
 * @param {string} [statusFilter="InProgress"] - The status filter for the jobs.
 * @returns {Promise<Object|null>} - The response from the ListAsyncInvokesCommand if successful, otherwise null.
 */
async function listJobStatuses(maxResults = 10, statusFilter = "InProgress") {
    try {
        const command = new ListAsyncInvokesCommand({
            maxResults: maxResults,
            statusEquals: statusFilter,
        });
        
        const invocation = await bedrockRuntime.send(command);
        
        logger.info("Invocation Jobs:");
        logger.info(JSON.stringify(invocation, null, 2));
        return invocation;
    } catch (error) {
        logger.error(`Error listing jobs: ${error}`);
        return null;
    }
}

/**
 * Monitors recent video generation jobs and downloads the videos.
 * @param {number} [durationHours=1] - The duration in hours to monitor jobs for.
 * @returns {Promise<void>} - A promise that resolves when the monitoring is complete.
 */
async function monitorRecentJobs(durationHours = 1) {
    const fromSubmitTime = new Date(Date.now() - durationHours * 60 * 60 * 1000);
    return await monitorAndDownloadVideos("output", fromSubmitTime);
}

/**
 * The main function to orchestrate video generation, status checking, and downloading.
 * @returns {Promise<void>} - A promise that resolves when the main function completes.
 */
async function main() {
    // Initialize the AWS session
    setupAwsSession();

    // Configuration
    const S3_BUCKET = "nova-videos"; // Change this to your unique bucket name
    const VIDEO_PROMPT = "Closeup of a large seashell in the sand, gentle waves flow around the shell. Camera zoom in.";
    const MODEL_ID = "amazon.nova-reel-v1:0";

    // Generate video
    const invocationArn = await generateVideo(S3_BUCKET, VIDEO_PROMPT, MODEL_ID);
    if (!invocationArn) {
        logger.error("Failed to start video generation");
        process.exit(1);
    }

    // Check initial status
    logger.info("\nChecking initial job status...");
    await checkJobStatus(invocationArn);

    // List all in-progress jobs
    logger.info("\nListing all in-progress jobs...");
    await listJobStatuses();

    // Monitor and download the video
    logger.info("\nMonitoring job and waiting for completion...");
    const localFilePath = await monitorAndDownloadVideo(invocationArn, "output");

    if (localFilePath) {
        logger.info(`\nVideo successfully generated and downloaded to: ${localFilePath}`);
    } else {
        logger.info("\nFailed to generate or download video");
    }
}

// Self-executing async function
(async () => {
    try {
        await main();
    } catch (err) {
        console.error('Fatal error:', err.message);
        process.exit(1);
    }
})();

File: multimodal-generation/video-generation/python/01_text_to_video_generation.py
#!/usr/bin/env python3

import boto3
import logging
import json
import random
import amazon_video_util
from datetime import datetime, timedelta, timezone

# Configure logging
logging.basicConfig(level=logging.DEBUG, format='[%(levelname)s] %(message)s')
logger = logging.getLogger(__name__)

# Constants
DEFAULT_MODEL_ID = "amazon.nova-reel-v1:0"

def setup_aws_session(region="us-east-1"):
    """Set up AWS session with default region"""
    boto3.setup_default_session(region_name=region)
    logger.info("AWS SDK session defaults have been set.")


def generate_video(s3_destination_bucket, video_prompt, model_id=DEFAULT_MODEL_ID):
    """
    Generate a video using the provided prompt.
    
    Args:
        s3_destination_bucket (str): The S3 bucket where the video will be stored
        video_prompt (str): Text prompt describing the desired video
    """
    # Set up the S3 client
    s3_client = boto3.client("s3")

    # Create the S3 bucket
    s3_client.create_bucket(Bucket=s3_destination_bucket)

    # Create the Bedrock Runtime client
    bedrock_runtime = boto3.client("bedrock-runtime")

    model_input = {
        "taskType": "TEXT_VIDEO",
        "textToVideoParams": {
            "text": video_prompt,
        },
        "videoGenerationConfig": {
            "durationSeconds": 6,  # 6 is the only supported value currently
            "fps": 24,  # 24 is the only supported value currently
            "dimension": "1280x720",  # "1280x720" is the only supported value currently
            "seed": random.randint(
                0, 2147483648
            ),  # A random seed guarantees we'll get a different result each time this code runs
        },
    }

    try:
        # Start the asynchronous video generation job
        invocation = bedrock_runtime.start_async_invoke(
            modelId=model_id,
            modelInput=model_input,
            outputDataConfig={"s3OutputDataConfig": {"s3Uri": f"s3://{s3_destination_bucket}"}},
        )

        # Store the invocation ARN
        invocation_arn = invocation["invocationArn"]

        # Pretty print the response JSON
        logger.info("\nResponse:")
        logger.info(json.dumps(invocation, indent=2, default=str))

        # Save the invocation details for later reference
        amazon_video_util.save_invocation_info(invocation, model_input)

        return invocation_arn

    except Exception as e:
        logger.error(e)
        return None

def check_job_status(invocation_arn):
    """Check status of a specific job using get_async_invoke()"""
    bedrock_runtime = boto3.client("bedrock-runtime")
    try:
        response = bedrock_runtime.get_async_invoke(
            invocationArn=invocation_arn
        )
        
        status = response["status"]
        logger.info(f"Status: {status}")
        logger.info("\nFull response:")
        logger.info(json.dumps(response, indent=2, default=str))
        return response
    except Exception as e:
        logger.error(f"Error checking job status: {e}")
        return None

def list_job_statuses(max_results=10, status_filter="InProgress"):
    """List all video generation jobs with optional filtering"""
    bedrock_runtime = boto3.client("bedrock-runtime")
    try:
        invocation = bedrock_runtime.list_async_invokes(
            maxResults=max_results,
            statusEquals=status_filter,
        )
        
        logger.info("Invocation Jobs:")
        logger.info(json.dumps(invocation, indent=2, default=str))
        return invocation
    except Exception as e:
        logger.error(f"Error listing jobs: {e}")
        return None

def monitor_recent_jobs(duration_hours=1):
    """Monitor and download videos from the past N hours"""
    from_submit_time = datetime.now(timezone.utc) - timedelta(hours=duration_hours)
    return amazon_video_util.monitor_and_download_videos("output", submit_time_after=from_submit_time)


def main():
    # Initialize the AWS session
    setup_aws_session()

    # Configuration
    S3_BUCKET = "nova-videos"  # Change this to your unique bucket name
    VIDEO_PROMPT = "Closeup of a large seashell in the sand, gentle waves flow around the shell. Camera zoom in."
    MODEL_ID = "amazon.nova-reel-v1:0"

    # Generate video
    invocation_arn = generate_video(S3_BUCKET, VIDEO_PROMPT, MODEL_ID)
    if not invocation_arn:
        logger.error("Failed to start video generation")
        exit(1)

    # Check initial status
    logger.info("\nChecking initial job status...")
    check_job_status(invocation_arn)

    # List all in-progress jobs
    logger.info("\nListing all in-progress jobs...")
    list_job_statuses()

    # Monitor and download the video
    logger.info("\nMonitoring job and waiting for completion...")
    local_file_path = amazon_video_util.monitor_and_download_video(invocation_arn, "output")

    if local_file_path:
        logger.info(f"\nVideo successfully generated and downloaded to: {local_file_path}")
    else:
        logger.info("\nFailed to generate or download video")

if __name__ == "__main__":
    main()